common/persistence/cassandra/metadata_store.go

```diff
@@ -161,22 +168,35 @@ func (m *MetadataStore) CreateNamespaceInV2Table(
 
 	previous := make(map[string]interface{})
 	applied, iter, err := m.session.MapExecuteBatchCAS(batch, previous)
+	deleteOrphanNamespace := func() {
+		// Delete namespace from `namespaces_by_id`
+		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
+			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
+		}
+	}
+
 	if err != nil {
 		return nil, serviceerror.NewUnavailable(fmt.Sprintf("CreateNamespace operation failed. Inserting into namespaces table. Error: %v", err))
 	}
+
 	defer func() { _ = iter.Close() }()
 
 	if !applied {
-		// Namespace already exist.  Delete orphan namespace record before returning back to user
-		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
-			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
-		}
+		if id, ok := previous["id"]; ok {
+			id = gocql.UUIDToString(id)
+			if id != request.ID {
+				// Namespace with the given name already exists with a different ID.
+				// Delete orphan namespace record before returning back to user
+				deleteOrphanNamespace()
```

Comment:
Since we are calling this method if an entry with the same `id` and `name` exists in `namespace_by_id` table, if we find out that an entry with same `name` but different `id` exists, it means a new row was added to `namespace_by_id` table and should be removed, otherwise having an existing row is expected.